src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs

```diff
@@ -280,27 +271,29 @@ public virtual string ReadString()
                 return string.Empty;
             }
 
-            _charBytes ??= new byte[MaxCharBytesSize];
+            _decoder ??= _encoding.GetDecoder();
             _charBuffer ??= new char[_maxCharsSize];
 
+            Span<byte> charBytes = stackalloc byte[MaxCharBytesSize];
+
             StringBuilder? sb = null;
             do
             {
                 readLength = ((stringLength - currPos) > MaxCharBytesSize) ? MaxCharBytesSize : (stringLength - currPos);
 
-                n = _stream.Read(_charBytes, 0, readLength);
+                n = _stream.Read(charBytes[..readLength]);
                 if (n == 0)
                 {
                     ThrowHelper.ThrowEndOfFileException();
                 }
 
-                charsRead = _decoder.GetChars(_charBytes, 0, n, _charBuffer, 0);
-
                 if (currPos == 0 && n == stringLength)
                 {
-                    return new string(_charBuffer, 0, charsRead);
+                    return _encoding.GetString(charBytes[..n]);
```

Comment:
it would be great to benchmark this test case as well

Reply:
Done in dotnet/performance#3114.